torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1396,16 +1448,19 @@ def _gather_orig_param_state(
     ]
     dist.all_gather_object(object_list, state_objects)
     orig_state: Dict[str, Any] = {}
-    for state in object_list:
+    device = torch.device("cuda", torch.cuda.current_device())
+    for idx, state in enumerate(object_list):
         for state_name, value in state.items():
             curr_value = orig_state.get(state_name, [])
             if torch.is_tensor(value):
                 if value.dim() > 0:
-                    curr_value.append(value)
-                elif torch.is_tensor(curr_value):
-                    assert torch.equal(curr_value, value)
-                else:
-                    orig_state[state_name] = value
+                    curr_value.append(value.to(device))
```

Comment:
This is not in scope for this PR, but I wonder if this will break CPU offloading, which runs the optimizer step on CPU.